blockchain/fullblocktests/generate.go

```diff
@@ -2044,6 +2068,55 @@ func Generate(includeLargeReorg bool) (tests [][]TestInstance, err error) {
 	}
 	accepted()
 
+	// Create a chain where the utxo created in b82a is spent in b83a.
```

Comment:
If I check out only this commit, then run the tests (expecting a failing test case), I get this panic: 
```
    fullblocks_test.go:220: Testing block b82 (hash 42246d9ea2cef3bac886bf25591884535d4206da1e5a3ced9317c4015a4ebec9, height 129)
    fullblocks_test.go:339: Testing tip for block b83a (hash 119f8757e953251625d83e585718acc4f43fa0fd12a657e885e08b1294158e5f, height 130)
    fullblocks_test.go:220: Testing block b83 (hash 4f95ef6bb94d5bcec3700a619caedb103bf88c8a37fe0c36da9c8a0ef5cc6103, height 130)
    fullblocks_test.go:339: Testing tip for block b83a (hash 119f8757e953251625d83e585718acc4f43fa0fd12a657e885e08b1294158e5f, height 130)
    fullblocks_test.go:220: Testing block b84 (hash 0599c9a4b9f468b3d30a2734e7da26830cbc67faeab1b65d40adedf5f3dde3fd, height 131)
    fullblocks_test.go:198: unexpected utxo 91f9eaef0b25c7905841d855c70753f51cc71d714f9ffd8f0a0d88a248653f9a:0 to exist but it's not nil and not spent
    fullblocks_test.go:200: unexpected utxo 91f9eaef0b25c7905841d855c70753f51cc71d714f9ffd8f0a0d88a248653f9a:0 exists but it's not nil and not spent


fatal error: sync: Unlock of unlocked RWMutex

goroutine 1628 [running]:
sync.fatal({0x1002b6e4c?, 0x1400009a028?})
	/Users/roasbeef/go/src/runtime/panic.go:1061 +0x20
sync.(*RWMutex).Unlock(0x14005dea07c)
	/Users/roasbeef/go/src/sync/rwmutex.go:209 +0x74
runtime.Goexit()
	/Users/roasbeef/go/src/runtime/panic.go:523 +0x14c
testing.(*common).FailNow(0x140013a0d00)
	/Users/roasbeef/go/src/testing/testing.go:999 +0x48
testing.(*common).Fatalf(0x140013a0d00, {0x1002bfb90?, 0x8390e4f3e79edc26?}, {0x14002db14b8?, 0x14000000000?, 0x0?})
	/Users/roasbeef/go/src/testing/testing.go:1083 +0x64
github.com/btcsuite/btcd/blockchain_test.TestFullBlocks.func1.2(0x100434f20?)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/fullblocks_test.go:200 +0x1dc
github.com/btcsuite/btcd/blockchain.(*BlockChain).sendNotification(0x14005dea000, 0x2, {0x100467800?, 0x14006b97ab0})
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/notifications.go:78 +0xf8
github.com/btcsuite/btcd/blockchain.(*BlockChain).disconnectBlock(0x14005dea000, 0x14001213d80, 0x14006b97ab0, 0x14005138b10)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/chain.go:857 +0x4a0
github.com/btcsuite/btcd/blockchain.(*BlockChain).reorganizeChain(0x14005dea000, 0x14003d7bf80, 0x14003d7bf50)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/chain.go:1093 +0xc40
github.com/btcsuite/btcd/blockchain.(*BlockChain).connectBestChain(0x14005dea000, 0x14001213f80, 0x14006b97880, 0x0)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/chain.go:1288 +0x670
github.com/btcsuite/btcd/blockchain.(*BlockChain).maybeAcceptBlock(0x14005dea000, 0x14006b97880, 0x489868?)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/accept.go:79 +0x160
github.com/btcsuite/btcd/blockchain.(*BlockChain).ProcessBlock(0x14005dea000, 0x14006b97880, 0x0)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/process.go:228 +0x258
github.com/btcsuite/btcd/blockchain_test.TestFullBlocks.func2({{0x1002aaa45, 0x3}, 0x1400116c180, 0x83, 0x1, 0x0})
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/fullblocks_test.go:223 +0x138
github.com/btcsuite/btcd/blockchain_test.TestFullBlocks(0x140013a0d00)
	/Users/roasbeef/gocode/src/github.com/btcsuite/btcd/blockchain/fullblocks_test.go:358 +0x3d0
testing.tRunner(0x140013a0d00, 0x100486208)
	/Users/roasbeef/go/src/testing/testing.go:1595 +0xe8
created by testing.(*T).Run in goroutine 1
	/Users/roasbeef/go/src/testing/testing.go:1648 +0x33c
```

Reply:
Ah yeah that's supposed to happen for any errors that happen during `b.sendNotification()` because the `t.Fatal(err)` happens first and then the `b.chainLock.Lock()` never gets called.

https://github.com/btcsuite/btcd/blob/13152b35e191385a874294a9dbc902e48b1d71b0/blockchain/chain.go#L853-L858

The `defer` in `ProcessBlock` will still get triggered so the runtime complains that you've unlocked an unlocked lock.

https://github.com/btcsuite/btcd/blob/13152b35e191385a874294a9dbc902e48b1d71b0/blockchain/process.go#L142-L144

I guess we should do the below. I'll make a separate PR for this as well.

```diff
diff --git a/blockchain/chain.go b/blockchain/chain.go
index 60420022..416ed929 100644
--- a/blockchain/chain.go
+++ b/blockchain/chain.go
@@ -730,9 +730,11 @@ func (b *BlockChain) connectBlock(node *blockNode, block *btcutil.Block,
        // Notify the caller that the block was connected to the main chain.
        // The caller would typically want to react with actions such as
        // updating wallets.
-       b.chainLock.Unlock()
-       b.sendNotification(NTBlockConnected, block)
-       b.chainLock.Lock()
+       func() {
+               b.chainLock.Unlock()
+               defer b.chainLock.Lock()
+               b.sendNotification(NTBlockConnected, block)
+       }()
 
        // Since we may have changed the UTXO cache, we make sure it didn't exceed its
        // maximum size.  If we're pruned and have flushed already, this will be a no-op.
@@ -853,9 +855,11 @@ func (b *BlockChain) disconnectBlock(node *blockNode, block *btcutil.Block, view
        // Notify the caller that the block was disconnected from the main
        // chain.  The caller would typically want to react with actions such as
        // updating wallets.
-       b.chainLock.Unlock()
-       b.sendNotification(NTBlockDisconnected, block)
-       b.chainLock.Lock()
+       func() {
+               b.chainLock.Unlock()
+               defer b.chainLock.Lock()
+               b.sendNotification(NTBlockDisconnected, block)
+       }()
 
        return nil
}
```

Reply:
https://github.com/btcsuite/btcd/pull/2128